stablehlo/reference/Ops.cpp

```diff
@@ -102,6 +102,123 @@ Tensor evalClampOp(const Tensor &min, const Tensor &operand, const Tensor &max,
   return result;
 }
 
+Tensor evalCompareOp(const Tensor &lhs, const Tensor &rhs,
+                     ComparisonDirection comparisonDirection,
+                     std::optional<ComparisonType> compareType,
+                     TensorType resultType) {
+  Tensor result(resultType);
+  auto elementTy = result.getType().dyn_cast<TensorType>().getElementType();
+  auto comparisonType = ComparisonType::NOTYPE;
+  if (compareType && *compareType != ComparisonType::NOTYPE) {
+    comparisonType = *compareType;
+  } else if (isSupportedUnsignedIntegerType(elementTy) ||
+             isSupportedBooleanType(elementTy)) {
+    comparisonType = ComparisonType::UNSIGNED;
+  } else if (isSupportedSignedIntegerType(elementTy)) {
+    comparisonType = ComparisonType::SIGNED;
+  } else if (isSupportedFloatType(elementTy) ||
+             isSupportedComplexType(elementTy)) {
+    comparisonType = ComparisonType::FLOAT;
+  } else {
+    report_fatal_error(invalidArgument("Unsupported element type: %s",
+                                       debugString(elementTy).c_str()));
+  }
+
+  for (auto it = result.index_begin(); it != result.index_end(); ++it) {
+    bool cmpResult = false;
+    switch (comparisonDirection) {
+      case ComparisonDirection::EQ: {
+        if (comparisonType == ComparisonType::TOTALORDER) {
+          auto lhsFloat = lhs.get(*it).getFloatValue();
+          auto rhsFloat = rhs.get(*it).getFloatValue();
+          cmpResult = lhs.get(*it).getFloatValue().bitcastToAPInt() ==
+                      rhs.get(*it).getFloatValue().bitcastToAPInt();
+        } else {
+          cmpResult = lhs.get(*it) == rhs.get(*it);
+        }
+        break;
+      }
+      case ComparisonDirection::NE: {
+        if (comparisonType == ComparisonType::TOTALORDER) {
+          auto lhsFloat = lhs.get(*it).getFloatValue();
+          auto rhsFloat = rhs.get(*it).getFloatValue();
+          cmpResult = lhs.get(*it).getFloatValue().bitcastToAPInt() !=
+                      rhs.get(*it).getFloatValue().bitcastToAPInt();
+        } else {
+          cmpResult = lhs.get(*it) != rhs.get(*it);
+        }
+        break;
+      }
+      case ComparisonDirection::GE: {
+        if (comparisonType == ComparisonType::TOTALORDER) {
+          auto lhsFloat = lhs.get(*it).getFloatValue();
+          auto rhsFloat = rhs.get(*it).getFloatValue();
+          if (lhsFloat.isNegative() ^ rhsFloat.isNegative())
+            cmpResult = rhsFloat.isNegative();
+          else
+            cmpResult =
+                lhsFloat.isNegative()
+                    ? lhsFloat.bitcastToAPInt().ule(rhsFloat.bitcastToAPInt())
+                    : lhsFloat.bitcastToAPInt().uge(rhsFloat.bitcastToAPInt());
```

Comment:
Also, is anyone using `TOTALORDER`? I skimmed the usual suspects, and I don't think they do. 

Given that, I think that we should just skip implementing it in the interpreter and (in this PR) update the spec along the lines of "For floating-point element types and `compare_type = TOTALORDER`, the op uses the combination of `totalOrder` and `compareQuietEqual` operations from IEEE-754. This feature appears to be unused, so in the future, we are planning to remove it ([#584](https://github.com/openxla/stablehlo/issues/584))"?

Also, I've just moved the corresponding ticket from New Features (P1) to Frontend Contract (P0) because the presence of this feature seems to have non-trivial poitential for confusion, so I think we should resolve it by v1.0.

Reply:
#584 links to HLO code that says this is deprecated: https://github.com/tensorflow/tensorflow/blob/3451f5d39ea8add4c254acc9615d62ff589259b8/tensorflow/compiler/xla/comparison_util.h#L71-L79
```
  // (DEPRECATED) Represents the type of comparison. Prefer xla::PrimitiveType
  // and Comparison::Order, since there are multiple floating point
  // representations that support total ordering.
  enum class [[deprecated("Use PrimitiveType and Order")]] Type : uint8_t{
      kFloat,
      kFloatTotalOrder,
      kSigned,
      kUnsigned,
  };
```
but a few lines above that, there is another enum class introduced to replace this, and indeed it is being used in [HLO](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/comparison_util.cc#L55).

https://github.com/tensorflow/tensorflow/blob/3451f5d39ea8add4c254acc9615d62ff589259b8/tensorflow/compiler/xla/comparison_util.h#L54-L59
```
  // Represents the ordering of the comparison.
  enum class Order : uint8_t {
    // https://en.wikipedia.org/wiki/Total_order
    kTotal,
    // https://en.wikipedia.org/wiki/Partially_ordered_set
    kPartial,
  };
```

Reply:
It is evident that `Order::kTotal` seems to be used in HLO. However, is it used by frontends? That's what would matter for including or not including this in StableHLO.

Reply:
I looked into this recently, and it looks like none of the frontends are using it specifically in CompareOp (e.g. they may use total order in, say, sort op by default) or is a backend implementation detail.